torch/testing/_asserts.py

```diff
@@ -102,6 +102,36 @@ def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Excepti
     return wrapper
 
 
+def _check_sparse_indices_and_values_individually(
+    check_tensors: Callable[..., Optional[Exception]]
+) -> Callable[..., Optional[Exception]]:
+    """Decorates dense tensor check functions to handle indices and values of sparse tensors individually.
+
+    If the inputs are not sparse, this decorator is a no-op.
+
+    Args:
+        check_tensors (Callable[[Tensor, Tensor], Optional[Exception]]): Tensor check function for dense
+        tensors.
+    """
+    @functools.wraps(check_tensors)
+    def wrapper(actual: Tensor, expected: Tensor, **kwargs: Any) -> Optional[Exception]:
+        if not actual.is_sparse:
+            return check_tensors(actual, expected, **kwargs)
+
+        actual = actual.coalesce()
```

Comment:
if actual is sparse but expected isn't won't this throw a runtime error in the assert function?

Reply:
Won't happen, since this only happens after `_check_attributes_equal`. See https://github.com/pytorch/pytorch/pull/58844#discussion_r642016652.